tensortrade/actions/discrete_actions.py

```diff
@@ -57,32 +57,46 @@ def dtype(self, dtype: DTypeString):
 
+    @property
+    def n_splits(self) -> int:
+        return self.n_actions / len(TradeType)
+
+    def _get_trade_type(self, action:TradeActionUnion) -> TradeType:
+        """calculates the trade type based off of the action value passed
+        params:
+            action: the action between the range of the n_actions
+        returns:
+            TradeType
+        """
+        return TradeType(action % len(TradeType))
+
+    def _get_trade_amount(self, action:TradeActionUnion) -> float:
+        amount = int(action / len(TradeType)) * float(1 / self.n_splits) + (1 / self.n_splits)
+        return amount
+
     def get_trade(self, action: TradeActionUnion) -> Trade:
         """The trade type is determined by `action % len(TradeType)`, and the trade amount is determined by the multiplicity of the action.
 
-        For example, 1 = LIMIT_BUY|0.25, 2 = MARKET_BUY|0.25, 6 = LIMIT_BUY|0.5, 7 = MARKET_BUY|0.5, etc.
+        For example, 1 = LIMIT_BUY|0.25, 2 = MARKET_BUY|0.25, 6 = LIMIT_BUY|0.5, 7 = MARKE
+        T_BUY|0.5, etc.
         """
-        n_splits = self.n_actions / len(TradeType)
-        trade_type = TradeType(action % len(TradeType))
-        trade_amount = int(action / len(TradeType)) * float(1 / n_splits) + (1 / n_splits)
 
-        current_price = self._exchange.current_price(symbol=self._instrument)
+        trade_type = self._get_trade_type(action)
+        trade_amount = self._get_trade_amount(action)
+
         base_precision = self._exchange.base_precision
         instrument_precision = self._exchange.instrument_precision
 
-        amount = self._exchange.instrument_balance(self._instrument)
-        price = current_price
+        amount_held = self._exchange.instrument_balance(self._instrument)
+        current_price = self._exchange.current_price(symbol=self._instrument)
 
-        if trade_type is TradeType.MARKET_BUY or trade_type is TradeType.LIMIT_BUY:
-            price_adjustment = 1 + (self.max_allowed_slippage_percent / 100)
-            price = max(round(current_price * price_adjustment, base_precision), base_precision)
-            amount = round(self._exchange.balance * 0.99 *
-                           trade_amount / price, instrument_precision)
+        balance = self._exchange.balance
 
-        elif trade_type is TradeType.MARKET_SELL or trade_type is TradeType.LIMIT_SELL:
-            price_adjustment = 1 - (self.max_allowed_slippage_percent / 100)
-            price = round(current_price * price_adjustment, base_precision)
-            amount_held = self._exchange.portfolio.get(self._instrument, 0)
-            amount = round(amount_held * trade_amount, instrument_precision)
+        if trade_type.is_buy:
+            # as an aditional barrier to overflowing the balance,
+            # we reserve 2% of our total balance for unseen slip and comissions.
+            trade_amount = round(balance * 0.98 * trade_amount / current_price, instrument_precision)
```

Comment:
This hard-coded `0.98` seems like a hack - is there a better way of doing this? 

The slippage would only come into account for MARKET_BUY, and as such the amount bought would be slipped, not the total amount traded. I believe the `* 0.98` is unnecessary.

Reply:
The 0.98 was originally 0.99 to prevent overspending. At some point as a quick fix I changed it to 0.98 to give it more headroom. With the current changes the withholding is either unnecessary, or if it is, should be configurable.